extension/parquet/parquet_metadata.cpp

```diff
@@ -813,133 +957,143 @@ static unique_ptr<FunctionData> ParquetMetaDataBind(ClientContext &context, Tabl
-		auto &bloom_probe_bind_data = input.bind_data->Cast<ParquetBloomProbeBindData>();
-		result->ExecuteBloomProbe(context, bind_data.return_types, bind_data.file_list->GetFirstFile(),
-		                          bloom_probe_bind_data.probe_column_name, bloom_probe_bind_data.probe_constant);
-		break;
+		const auto &probe_bind_data = static_cast<const ParquetBloomProbeBindData &>(bind_data);
+		return make_uniq<ParquetBloomProbeProcessor>(context, file_path, probe_bind_data.probe_column_name,
+		                                             probe_bind_data.probe_constant);
 	}
 	default:
 		throw InternalException("Unsupported ParquetMetadataOperatorType");
 	}
-
-	return std::move(result);
-}
-
-template <ParquetMetadataOperatorType TYPE>
-static void ParquetMetaDataImplementation(ClientContext &context, TableFunctionInput &data_p, DataChunk &output) {
-	auto &data = data_p.global_state->Cast<ParquetMetaDataOperatorData>();
-	auto &bind_data = data_p.bind_data->Cast<ParquetMetaDataBindData>();
-
-	while (true) {
-		if (!data.collection.Scan(data.scan_state, output)) {
-
-			// Try get next file
-			if (!bind_data.file_list->Scan(data.file_list_scan, data.current_file)) {
-				return;
-			}
-
-			switch (TYPE) {
-			case ParquetMetadataOperatorType::SCHEMA:
-				data.LoadSchemaData(context, bind_data.return_types, data.current_file);
-				break;
-			case ParquetMetadataOperatorType::META_DATA:
-				data.LoadRowGroupMetadata(context, bind_data.return_types, data.current_file);
-				break;
-			case ParquetMetadataOperatorType::KEY_VALUE_META_DATA:
-				data.LoadKeyValueMetaData(context, bind_data.return_types, data.current_file);
-				break;
-			case ParquetMetadataOperatorType::FILE_META_DATA:
-				data.LoadFileMetaData(context, bind_data.return_types, data.current_file);
-				break;
-			case ParquetMetadataOperatorType::BLOOM_PROBE: {
-				auto &bloom_probe_bind_data = data_p.bind_data->Cast<ParquetBloomProbeBindData>();
-				data.ExecuteBloomProbe(context, bind_data.return_types, data.current_file,
-				                       bloom_probe_bind_data.probe_column_name, bloom_probe_bind_data.probe_constant);
-				break;
-			}
-			default:
-				throw InternalException("Unsupported ParquetMetadataOperatorType");
-			}
-			continue;
-		}
-		if (output.size() != 0) {
-			return;
-		}
-	}
 }
 
 ParquetMetaDataFunction::ParquetMetaDataFunction()
     : TableFunction("parquet_metadata", {LogicalType::VARCHAR},
-                    ParquetMetaDataImplementation<ParquetMetadataOperatorType::META_DATA>,
-                    ParquetMetaDataBind<ParquetMetadataOperatorType::META_DATA>,
-                    ParquetMetaDataInit<ParquetMetadataOperatorType::META_DATA>) {
+                    ParquetMetaDataOperator<ParquetMetadataOperatorType::META_DATA>::Function,
+                    ParquetMetaDataOperator<ParquetMetadataOperatorType::META_DATA>::Bind,
+                    ParquetMetaDataOperator<ParquetMetadataOperatorType::META_DATA>::InitGlobal,
```

Comment:
This templating seems to only really be required in the `Bind` and `Function` - maybe we can move the template from the class to these functions instead?